rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,221 @@
+    if let Some(iterator) = iterator {
+        let s = match iterator.peek() {
+            // TODO: Replace with fallible `to_string` when available.
+            Some(item) => item.to_string() + "\0",
+            None => "\0".to_string(),
+        };
+        // SAFETY: Calling a C function. `s` is guaranteed to be null terminated
+        // because we explicitly constructed it just above.
+        unsafe {
+            bindings::seq_puts(m, s.as_ptr() as *const u8 as *const c_types::c_char);
+        }
+    }
+    0
+}
+
+extern "C" fn start_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    // SAFETY: This function will be called by opening a proc file generated
+    // by a call to `proc_create::<T>`, so `m` can be converted using
+    // `SeqFile<T>`.
+    let arg = unsafe { &*SeqFile::<T>::convert(m) };
+    // SAFETY: The caller guarantees that `pos` points to a valid `loff_t`.
+    let pos = unsafe { *pos };
+    match T::start(arg) {
+        Some(mut wrapper) => {
+            for _ in 0..pos {
+                wrapper.as_mut().next();
+            }
+            Box::into_raw(wrapper) as *mut c_types::c_void
+        }
+        None => ptr::null_mut(),
+    }
+}
+
+struct SeqFileOperationsVTable<T>(PhantomData<T>);
+
+impl<T: SeqOperations> SeqFileOperationsVTable<T> {
+    const VTABLE: bindings::seq_operations = bindings::seq_operations {
+        start: Some(start_callback::<T>),
+        stop: Some(stop_callback::<T>),
+        next: Some(next_callback::<T>),
+        show: Some(show_callback::<T>),
+    };
+}
+
+/// A `seq_file` referencing data of type `T`.
+pub struct SeqFile<T> {
+    seq_ops: bindings::seq_operations,
+    _pin: PhantomPinned,
+    dir_entry: *mut bindings::proc_dir_entry,
+    context: T,
+}
+
+impl<T> Drop for SeqFile<T> {
+    fn drop(&mut self) {
+        if !self.dir_entry.is_null() {
+            // SAFETY: Calling a C function. `dir_entry` is a valid pointer to a
+            // `proc_dir_entry` because if it is not null it was created by a
+            // call to `bindings::proc_create_seq_private`.
+            unsafe {
+                bindings::proc_remove(self.dir_entry);
+            }
+        }
+    }
+}
+
+impl<T> SeqFile<T> {
+    fn new(seq_ops: bindings::seq_operations, context: T) -> KernelResult<Box<Self>> {
+        Ok(Box::try_new(SeqFile {
+            seq_ops,
+            _pin: PhantomPinned,
+            dir_entry: ptr::null_mut::<bindings::proc_dir_entry>(),
+            context,
+        })?)
+    }
+
+    /// Retrieve the [`SeqFile::context`] associated with a [`bindings::seq_file`].
+    ///
+    /// # Safety
+    ///
+    /// `m` must have been created from a proc file generated by calling
+    /// `proc_create::<T>`.
+    unsafe fn convert(m: *mut bindings::seq_file) -> *const T {
+        let reg = crate::container_of!((*m).op, Self, seq_ops);
```

Comment:
You should use `(*m).private` to save state. This way `op` can be constant shared by everyone.

Reply:
Had to make some changes to get this to work. I couldn't see how to stare the data in `(*m).private` using `proc_create_seq_private` because that always uses the default `seq_open` which doesn't give us the chance to mutate the `seq_file` created on open.

So I switched it to use `proc_create_data` which let's us use our own `open`.

Reply:
Thanks, got that working now.